src/forms/Typeahead.jsx

```diff
@@ -12,6 +12,56 @@ export const TA_ITEM_CLASSNAME = 'typeahead-item';
  * @module Typeahead
  */
 class Typeahead extends React.PureComponent {
+	constructor(props) {
+		super(props);
+		this.onItemMouseEnter = this.onItemMouseEnter.bind(this);
+		this.stateReducer = this.stateReducer.bind(this);
+
+		this.state = {
+			highlightedIndex: -1
+		};
+	}
+
+	onItemMouseEnter (i) {
+		this.setState({highlightedIndex: i});
+	}
+
+	stateReducer(state, changes) {
+		let highlightedIndexState;
+
+		switch (changes.type) {
+			case Downshift.stateChangeTypes.keyDownArrowDown:
+				highlightedIndexState =
+					parseInt(this.state.highlightedIndex + 1) >= this.props.items.length
+						?
+							0
+						:
+							parseInt(this.state.highlightedIndex+1);
+				this.setState({
+					highlightedIndex: highlightedIndexState
+				});
+				return {
```

Comment:
Because we're returning basically the same thing in each `case`, I'd recommend doing the return outside of the switch statement. Each `case` should end with `break;` instead. The `default` case can be removed for this statement.